lib/upgrade/etcd/conversionv1v3/hostendpoint.go

```diff
@@ -0,0 +1,139 @@
+// limitations under the License.
+
+package conversionv1v3
+
+import (
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apiv1 "github.com/projectcalico/libcalico-go/lib/apis/v1"
+	"github.com/projectcalico/libcalico-go/lib/apis/v1/unversioned"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/numorstring"
+)
+
+// HostEndpoint implements the Converter interface.
+type HostEndpoint struct{}
+
+// APIV1ToBackendV1 converts an APIv1 HostEndpoint structure to a KVPair containing a
+// backend HostEndpoint and HostEndpointKey.
+// This is part of the converter interface.
+func (_ HostEndpoint) APIV1ToBackendV1(a unversioned.Resource) (*model.KVPair, error) {
+	ah, ok := a.(apiv1.HostEndpoint)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid v1 HostEndpoint")
+	}
+	var ipv4Addrs []cnet.IP
+	var ipv6Addrs []cnet.IP
+	for _, ip := range ah.Spec.ExpectedIPs {
+		if ip.Version() == 4 {
+			ipv4Addrs = append(ipv4Addrs, ip)
+		} else {
+			ipv6Addrs = append(ipv6Addrs, ip)
+		}
+	}
+
+	var ports []model.EndpointPort
+	for _, port := range ah.Spec.Ports {
+		ports = append(ports, model.EndpointPort{
+			Name:     port.Name,
+			Protocol: port.Protocol,
+			Port:     port.Port,
+		})
+	}
+
+	d := model.KVPair{
+		Key: model.HostEndpointKey{
+			Hostname:   ah.Metadata.Node,
+			EndpointID: ah.Metadata.Name,
+		},
+		Value: &model.HostEndpoint{
+			Labels: ah.Metadata.Labels,
+
+			Name:              ah.Spec.InterfaceName,
+			ProfileIDs:        ah.Spec.Profiles,
+			ExpectedIPv4Addrs: ipv4Addrs,
+			ExpectedIPv6Addrs: ipv6Addrs,
+			Ports:             ports,
+		},
+	}
+
+	log.WithFields(log.Fields{
+		"v1HostEndpoint": a,
+		"KVPair":         d,
+	}).Debug("Converted HostEndpoint to KVPair")
+
+	return &d, nil
+}
+
+// BackendV1ToAPIV3 converts a KVPair containing a backend HostEndpoint and HostEndpointKey
+// to an APIv3 HostEndpoint structure.
+// This is part of the Converter interface.
+func (_ HostEndpoint) BackendV1ToAPIV3(d *model.KVPair) (Resource, error) {
+	bh, ok := d.Value.(*model.HostEndpoint)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid HostEndpoint value")
+	}
+	bk, ok := d.Key.(model.HostEndpointKey)
+	if !ok {
+		return nil, fmt.Errorf("key is not a valid HostEndpoint key")
+	}
+
+	var ips []string
+	for _, ip := range bh.ExpectedIPv4Addrs {
+		ips = append(ips, ip.String())
+	}
+	for _, ip := range bh.ExpectedIPv6Addrs {
+		ips = append(ips, ip.String())
+	}
+
+	var ports []apiv3.EndpointPort
+	for _, port := range bh.Ports {
+		ports = append(ports, apiv3.EndpointPort{
+			Name:     port.Name,
+			Protocol: numorstring.ProtocolV3FromProtocolV1(port.Protocol),
+			Port:     port.Port,
+		})
+	}
+
+	var pIds []string
+	for _, pId := range bh.ProfileIDs {
+		pIds = append(pIds, convertName(pId))
```

Comment:
Strictly speaking, we should have a `convertProfileName`.

We need this for `WorkloadEndpoints` and it probably should be done for `HostEndpoints` as well.

The main thing is that in v1, we create per-namespace profiles called `k8s_ns.<namespace>` and in v3 the equivalent is `kns.<namespace>`.  We need to make sure we're handling that transition here.

Reply:
CC @robbrockbank in case it's not on your radar.

Reply:
I think you need to use this function: https://github.com/projectcalico/libcalico-go/blob/master/lib/upgrade/etcd/conversionv1v3/workloadendpoint.go#L238 

As well as the name conversion.